### PR TITLE
first draft for supporting docker-compose v2

### PR DESCRIPTION
**Pull request by ggessner**
# Pull Request

## Disclaimer

I'm not a golang developer, so any code presented here is a collection of educated guesses as how to solve the problem.
With that out of the way, let's present the pull request.

## Problem
This pull request adresses [Issue231: firefly fails check for docker compose v2](https://github.com/hyperledger/firefly-cli/issues/231). The problem arises due to docker changing the syntax for docker-compose from `docker-compose` in version 1 to `docker compose` in version 2. There is always the workaround to require the standalone docker-compose in version 1.

## Fix

The fix is split into two parts, adressing the checking of the version and integrating the information into the stack manager that executes the docker commands via `RunDockerComposeCommand`. The information about which version syntax to use for the commands is carried via context.

### CheckDockerConfig
the function that checks for the prerequisites when executing docker-compose commands, `CheckDockerConfig` returns a tuple `(DockerComposeVersion, error)` with `DockerComposeVersion` as enum to differentiate between `None`, `ComposeV1` and `ComposeV2`. This value is stored in a context that is later transferred into the stack manager.
Unfortunately, cobra with version 1.4.0 does not provide the capability to allow context to be attached to a command.
Only from version 1.5.0, this feature has been merged, for reference: [support for SetContext](https://github.com/spf13/cobra/pull/1551).
The solution using cobra 1.5.0 concerns all commands which use both `PreRunE` and `RunE` while also requiring the docker compose version to be transferred as context between these two executions:

```go 
        PreRunE: func(cmd *cobra.Command, args []string) error {
		ctx := ...
		version, err := docker.CheckDockerConfig()
		ctx = context.WithValue(ctx, docker.ComposeVersion{}, version)
		cmd.SetContext(ctx)
		return err
	},
	RunE: func(cmd *cobra.Command, args []string) error {
                ...
		stackManager := stacks.NewStackManager(cmd.Context())
               ...
	},
```

The functions currently affected are defined in:
- `accounts_create.go`
- `accounts_list.go`
- `deploy_ethereum.go`
- `deploy_fabric`

No version update is required when simply combining both executions, similar to functions from:
- `info.go`
- `logs.go`
- `reset.go`
- `start.go`

```go 
        RunE: func(cmd *cobra.Command, args []string) error {
		ctx := ...
		version, err := docker.CheckDockerConfig()
		ctx = context.WithValue(ctx, docker.ComposeVersion{}, version)
                if err != nil {
                     return err
                }
		stackManager := stacks.NewStackManager(cmd.Context())
               ...
	},
```

### RunDockerComposeCommand

The second part of the changes affects the way the stack manager runs docker-compose functions. By retrieving the docker-compose version, either `docker-compose` or `docker compose` can be executed

## Possible Tweaks

- Instead of storing a version in the context, the command itself can be stored; when docker-compose in version 2 is detected, instead storing V1, the string `docker compose` is attached to the context with `ComposeCmd` as key. Consequently, the `RunDockerComposeCommand` function simply executes `exec.Command(ctx.Value(ComposeCmd{}), command...)`.

## Open Question

- Is the separation between `PreRunE` and `RunE` necessary, because currently only docker checks are executed in `PreRun`. Small commands do not even specify `PreRunE` at all.

### CODE REVIEW

**Comment by nguyer** [comment]
Okay, actually you can disregard my previous concern. That was an issue with the Compose Specification version (and going to V3), not the Docker Compose binary version.

File: internal/docker/docker_checks.go
Comment (nguyer) [reply internal/docker/docker_checks.go:52]: I also might suggest a `switch`/`case` statement here instead of several `else if`s. In Go, you don't actually have to `switch` "on" a variable.

**Review by nguyer — CHANGES_REQUESTED** [review]
This is a really great contribution, especially for a first contribution to the project! Thank you so much!

I think a couple of minor changes are needed, but after that, I would love to merge this and include it in the next release.

File: internal/docker/docker_checks.go
Comment (ggessner) [reply internal/docker/docker_checks.go:52]: Yes, having both installed is one of the tests missing. Although, handling that case is quite intriguing. My intuition would be to force the usage of the new one. Not out of disregard for stability but to ensure that users, without any settings or configuration, are set to use the newer future proof api docker provides. This in turn also leads to bugs being found that pertain to the switch between v1 and v2. Less radical is keeping v1 as the dominant, when both versions are present.

Comment (ggessner) [reply internal/docker/docker_checks.go:52]: I think another solution would be to keep the flow of the application and instead of using my first draft of convoluted checks to a more elegant solution;

```go
	// ckeck for docker-compose (v1) version
	dockerComposeCmd := exec.Command("docker-compose", "-v")
	_, err = dockerComposeCmd.Output()
	if err == nil {
		return ComposeV1, nil
	}

	// ckeck for docker-compose (V2) version
	dockerComposeCmd = exec.Command("docker compose", "version")
	_, err = dockerComposeCmd.Output()
	if err == nil {
		return ComposeV2, nil
	}

	return None, fmt.Errorf("an error occurred while running docker-compose. Is docker-compose installed on your computer?")
```

The checks are in sequence with the first check being the preferred if both were to be present.

Comment (ggessner) [reply internal/docker/docker_checks.go]: Another more complex solution is to allow the user to specify which docker-version he would like to use to invoke firefly-cli. But this would require a flag or parameter to be included in the available api of firefly-cli. I see this as a more extensive change and not in scope for this PR.

**Comment by nguyer** [comment]
This looks great now! The one last thing is to make sure you have sign-off on your commits. There are some details on how to do that here: https://github.com/hyperledger/firefly-cli/pull/232/checks?check_run_id=10537046269 It may be easiest to squash merge with one new commit with sign off. Sorry for the trouble, but this is just a requirement for all Linux Foundation repos.

**Comment by ggessner** [comment]
Hi @nguyer,

thanks for the advice. I squashed the commits and signed the resulting one.
Let me know if I made any errors.

**Comment by nguyer** [comment]
Sorry, it's not a _signed_ commit that we need, but rather _sign off_. You just need to add `-s` to your commit message and git will do it for you:

https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s

Here's a bit more info about it if you need it: https://wiki.linuxfoundation.org/dco

**Comment by ggessner** [comment]
That felt like an unnecessary mistake from my side.
It should be fixed now.

**Review by nguyer — APPROVED** [review]

**Comment by nguyer** [comment]
Looks great now! Thank you for contributing this!

**Pull request merged**
